File: auth_trustlines_demo.js
const xrpl = require("xrpl");

const TOKEN_CODE = "TES";
const TOKEN_AMOUNT = "100";

async function main() {
  const client = new xrpl.Client("wss://s.altnet.rippletest.net:51233");
  await client.connect();

  // 1. Create issuer and receiver wallets
  const issuer = (await client.fundWallet()).wallet;
  const receiver = (await client.fundWallet()).wallet;

  // 2. Enable Require Auth on issuer (must be first!)
  const setRequireAuthTx = {
    TransactionType: "AccountSet",
    Account: issuer.classicAddress,
    SetFlag: xrpl.AccountSetAsfFlags.asfRequireAuth
  };
  const requireauth_prepared = await client.autofill(setRequireAuthTx);
  const requireauth_signed = issuer.sign(requireauth_prepared);
  const requireauth_result = await client.submitAndWait(requireauth_signed.tx_blob);
  console.log("Require Auth enabled on issuer:", requireauth_result.result.meta.TransactionResult);

  // 3. Set Default Ripple on issuer
  const setDefaultRippleTx = {
    TransactionType: "AccountSet",
    Account: issuer.classicAddress,
    SetFlag: xrpl.AccountSetAsfFlags.asfDefaultRipple
  };
  const set_prepared = await client.autofill(setDefaultRippleTx);
  const set_signed = issuer.sign(set_prepared);
  const set_result = await client.submitAndWait(set_signed.tx_blob);
  console.log("Default Ripple enabled on issuer:", set_result.result.meta.TransactionResult);

  // 4. Set up trustline from receiver to issuer (should be unauthorized)
  const trustSetTx = {
    TransactionType: "TrustSet",
    Account: receiver.classicAddress,
    LimitAmount: {
      currency: TOKEN_CODE,
      issuer: issuer.classicAddress,
      value: TOKEN_AMOUNT,
    },
  };
  const ts_prepared = await client.autofill(trustSetTx);
  const ts_signed = receiver.sign(ts_prepared);
  const ts_result = await client.submitAndWait(ts_signed.tx_blob);
  console.log("Trustline set from receiver to issuer (unauthorized):", ts_result.result.meta.TransactionResult);

  // 5. Issuer tries to send tokens to receiver (should fail)
  const paymentTx = {
    TransactionType: "Payment",
    Account: issuer.classicAddress,
    Amount: {
      currency: TOKEN_CODE,
      value: TOKEN_AMOUNT,
      issuer: issuer.classicAddress,
    },
    Destination: receiver.classicAddress,
  };
  const pay_prepared = await client.autofill(paymentTx);
  const pay_signed = issuer.sign(pay_prepared);
  const pay_result = await client.submitAndWait(pay_signed.tx_blob);
  console.log("Issuer tried to send tokens to receiver (should fail):", pay_result.result.meta.TransactionResult);

  // 6. Issuer authorizes the trustline
  const authorizeTx = {
    TransactionType: "TrustSet",
    Account: issuer.classicAddress,
    LimitAmount: {
      currency: TOKEN_CODE,
      issuer: receiver.classicAddress,
      value: "0",
    },
    Flags: xrpl.TrustSetFlags.tfSetfAuth,
  };
  const auth_prepared = await client.autofill(authorizeTx);
  const auth_signed = issuer.sign(auth_prepared);
  const auth_result = await client.submitAndWait(auth_signed.tx_blob);
  console.log("Issuer authorized receiver's trustline:", auth_result.result.meta.TransactionResult);

  // 7. Issuer sends tokens to receiver (should succeed)
  const pay2_prepared = await client.autofill(paymentTx);
  const pay2_signed = issuer.sign(pay2_prepared);
  const pay2_result = await client.submitAndWait(pay2_signed.tx_blob);
  console.log("Issuer sent tokens to receiver (should succeed):", pay2_result.result.meta.TransactionResult);

  // 8. Issuer unwhitelists (removes authorization) from the trustline
  const unauthTx = {
    TransactionType: "TrustSet",
    Account: issuer.classicAddress,
    LimitAmount: {
      currency: TOKEN_CODE,
      issuer: receiver.classicAddress,
      value: "0",
    },
    Flags: xrpl.TrustSetFlags.tfClearAuth,
  };
  const unauth_prepared = await client.autofill(unauthTx);
  const unauth_signed = issuer.sign(unauth_prepared);
  const unauth_result = await client.submitAndWait(unauth_signed.tx_blob);
  console.log("Issuer removed authorization from receiver's trustline:", unauth_result.result.meta.TransactionResult);

  await client.disconnect();
}

main().catch(console.error); 

File: auth_trustlines_demo.py
import asyncio

from xrpl.asyncio.transaction import XRPLReliableSubmissionException
from xrpl.models.transactions import AccountSet, Payment, TrustSet
from xrpl.models.transactions.account_set import AccountSetAsfFlag
from xrpl.models.transactions.trust_set import TrustSetFlag

from xrpl_helpers import (
    fund_wallet,
    issued_currency_amount,
    submit_and_log,
    xrpl_client,
)

TOKEN_CODE = "TES"
TOKEN_AMOUNT = "100"


async def main() -> None:
    async with xrpl_client() as client:
        issuer = await fund_wallet(client, "issuer")
        receiver = await fund_wallet(client, "receiver")

        await submit_and_log(
            client,
            AccountSet(
                account=issuer.classic_address,
                set_flag=AccountSetAsfFlag.ASF_REQUIRE_AUTH,
            ),
            issuer,
            description="Require Auth enabled on issuer",
        )

        await submit_and_log(
            client,
            AccountSet(
                account=issuer.classic_address,
                set_flag=AccountSetAsfFlag.ASF_DEFAULT_RIPPLE,
            ),
            issuer,
            description="Default Ripple enabled on issuer",
        )

        await submit_and_log(
            client,
            TrustSet(
                account=receiver.classic_address,
                limit_amount=issued_currency_amount(
                    TOKEN_CODE, issuer.classic_address, TOKEN_AMOUNT
                ),
            ),
            receiver,
            description="Trustline set from receiver to issuer (unauthorized)",
        )

        try:
            await submit_and_log(
                client,
                Payment(
                    account=issuer.classic_address,
                    destination=receiver.classic_address,
                    amount=issued_currency_amount(
                        TOKEN_CODE, issuer.classic_address, TOKEN_AMOUNT
                    ),
                ),
                issuer,
                description="Issuer attempted to send tokens before authorization",
            )
        except XRPLReliableSubmissionException as exc:
            print(f"Expected failure before authorization: {exc}")

        await submit_and_log(
            client,
            TrustSet(
                account=issuer.classic_address,
                limit_amount=issued_currency_amount(
                    TOKEN_CODE, receiver.classic_address, "0"
                ),
                flags=TrustSetFlag.TF_SET_AUTH,
            ),
            issuer,
            description="Issuer authorized receiver's trustline",
        )

        await submit_and_log(
            client,
            Payment(
                account=issuer.classic_address,
                destination=receiver.classic_address,
                amount=issued_currency_amount(
                    TOKEN_CODE, issuer.classic_address, TOKEN_AMOUNT
                ),
            ),
            issuer,
            description="Issuer sent tokens after authorization",
        )

        print(
            "Authorization cannot be revoked once granted; tfClearAuth is not available."
        )


if __name__ == "__main__":
    asyncio.run(main())
